kerchunk/df.py

```diff
@@ -25,7 +28,182 @@ def _proc_raw(r):
     return r
 
 
+def get_variables(keys):
+    """Get list of variable names from references.
+
+    Parameters
+    ----------
+    keys : list of str
+        kerchunk references keys
+
+    Returns
+    -------
+    fields : list of str
+        List of variable names.
+    """
+    fields = []
+    for k in keys:
+        if "/" in k:
+            name, chunk = k.split("/")
+            if name not in fields:
+                fields.append(name)
+            else:
+                continue
+        else:
+            fields.append(k)
+    return fields
+
+
+def _normalize_json(json_obj):
+    """Normalize json representation as bytes
+
+    Parameters
+    ----------
+    json_obj : str, bytes, dict, list
+        JSON data for parquet file to be written.
+    """
+    if not isinstance(json_obj, str) and not isinstance(json_obj, bytes):
+        json_obj = ujson.dumps(json_obj)
+    if not isinstance(json_obj, bytes):
+        json_obj = json_obj.encode()
+    return json_obj
+
+
+def _write_json(fname, json_obj):
+    """Write references into a parquet file.
+
+    Parameters
+    ----------
+    fname : str
+        Output filename.
+    json_obj : str, bytes, dict, list
+        JSON data for parquet file to be written.
+    """
+    json_obj = _normalize_json(json_obj)
+    with open(fname, "wb") as f:
+        f.write(json_obj)
+
+
 def refs_to_dataframe(
+    refs,
+    url,
+    storage_options=None,
+    row_group_size=1000,
+    compression="zstd",
+    engine="fastparquet",
+    **kwargs,
+):
+    """Write references as a store of parquet files with multiple row groups.
+    The directory structure should mimic a normal zarr store but instead of standard chunk
+    keys, references are saved as parquet dataframes with multiple row groups.
+
+    Parameters
+    ----------
+    refs: str | dict
+        Location of a JSON file containing references or a reference set already loaded
+        into memory. It will get processed by the standard referenceFS, to normalise
+        any templates, etc., it might contain.
+    url: str
+        Location for the output, together with protocol. This must be a writable
+        directory.
+    storage_options: dict | None
+        Passed to fsspec when for writing the parquet.
+    row_group_size : int
+        Number of references to store in each reference file (default 1000)
+    compression : str
+        Compression information to pass to parquet engine, default is zstd.
+    engine : {'fastparquet', 'pyarrow'}
+        Library to use for writing parquet files.
+    **kwargs : dict
+        Additional keyword arguments passed to parquet engine of choice.
+    """
+    if not os.path.exists(url):
+        os.makedirs(url)
+    if "refs" in refs:
```

Comment:
Maybe there should be a better method to normalize refs here. The original `refs_to_dataframe` passes them into a separate `ReferenceFilesystem` to do this in one go, but I found that doing this is is really slow for large reference sets because (1) in the case a dict is passed, there is overhead from fsspec in tokenizing the input and (2) it needs to iterate through the whole reference set in case there are templates that need to be rendered. Wondering if we could/should add separate logic that skips doing this if there are no templates in the urls?